core/services/ocr2/plugins/ocr2vrf/coordinator/coordinator.go

```diff
@@ -739,3 +868,19 @@ func toGethLog(lg logpoller.Log) types.Log {
 		Index:       uint(lg.LogIndex),
 	}
 }
+
+// getBlockCacheKey returns a cache key for a requested block
+func getBlockCacheKey(blockNumber uint64, confDelay uint64) common.Hash {
+	var blockNumberBytes [8]byte
+	var confDelayBytes [8]byte
+
+	binary.BigEndian.PutUint64(blockNumberBytes[:], blockNumber)
+	binary.BigEndian.PutUint64(confDelayBytes[:], confDelay)
+
+	return common.BytesToHash(bytes.Join([][]byte{blockNumberBytes[:], confDelayBytes[:]}, nil))
+}
+
+// getBlockCacheKey returns a cache key for a requested callback
+func getCallbackCacheKey(requestID int64) common.Hash {
+	return common.BigToHash(big.NewInt(requestID))
```

Comment:
Given the following assumption:
- The Coordinator only caches items in `ReportWillBeTransmitted`

Ex. 1
A re-org that happens after two requests are made and after they are transmitted, that swaps the nonces of the original requests. This would be a deep re-org, and this re-org would invalidate the blockhash of the transmitted report associated with each callback. So neither's cache entry would be valid anymore (as soon as the log poller picks up the re-org), and they will both be re-transmitted. Subsequent valid cache attempts would override the old values.

Ex. 2
A re-org that happens after two requests are made, and after they are transmitted, that does not swap the original requests but invalidates the proofs for the callbacks' reports. The RecentBlockhash in the report for any of these callbacks will be invalidated, and the next round of ReportBlocks will pick up the callbacks as though they weren't cached.

Ex. 3
A re-org that happens after two requests are made, but before the report is transmitted, that swaps the nonces of the original requests. Any report containing re-orged/swapped requests will be rejected in `ReportWillBeTransmitted` due to an incorrect RecentBlockhash, and no items will be cached.

This was a good callout that I hadn't thought of. After looking into it, though, I think the current key set up will work.

Reply:
> So neither's cache entry would be valid anymore (as soon as the log poller picks up the re-org), and they will both be re-transmitted. Subsequent valid cache attempts would override the old values.

Can you explain this part? Would the cache entry be invalidated for the callback key (which is currently just request ID)? Where does that invalidation happen? The only eviction I see is in `EvictExpiredItems` which evicts expired items and doesn't invalidate e.g re-orged requests as far as I can tell. I think for regular `RandomnessRequest`s it might be fine because the next beacon output height and the conf delay will likely cause a different key, but it's still possible that after a re-org you'd have the same output height and conf delay, and the lookup [here](https://github.com/smartcontractkit/chainlink/blob/c8766ec087c627ee5175c275f40012bc9b5c5264/core/services/ocr2/plugins/ocr2vrf/coordinator/coordinator.go#L563) will yield a correct item. It seems similar for callbacks, whether they retain their request ID's or not.

It seems like it would be safest to include the blockhash in this cache key, that way we're ensured to have no collisions on re-orgs. Otherwise it seems like we'd be very likely to have collisions and as a result not re-transmit re-orged requests, unless I'm misunderstanding something.

Reply:
After chatting about this offline I understand why we don't have the blockhash in the key - the "invalidation" of items in the cache is implicit - basically, the interval cleaner will clean them up or the direct call to EvictExpiredItems in `ReportWillBeTransmitted`. The reason the blockhash is not in the key is that it will make it harder to look up from the cache. However, we _will_ re-fulfill re-orged requests because we end up checking whether the recent blockhash of the item in the cache is equal to the recent block hash as fetched from the chain.